### PR TITLE
Adding type in EncryptionKeyWrapMetadata and performing validation on partition key

### PR DESCRIPTION
**Pull request by simplynaveen20**
This PR contains 3 parts
1. Adding type in EncryptionKeyWrapMetadata constructor and doing validation of store provider name during CEK creation
2. Validation of partition key to avoid having it as encrypted field 
3. Exposing policyFormatVersion in ClientEncryptionPolicy

closes https://github.com/Azure/azure-sdk-for-java/issues/20376

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ClientEncryptionPolicy.java [hunk 1]
```diff
@@ -76,6 +79,33 @@ public ClientEncryptionPolicy() {
...
+     */
+    @Beta(value = Beta.SinceVersion.V4_16_0, warningText = Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
+    public int getPolicyFormatVersion() {
+        return policyFormatVersion;
+    }
+
+    void validatePartitionKeyPathsAreNotEncrypted(List<List<String>> partitionKeyPathTokens) {
+        checkNotNull(partitionKeyPathTokens, "partitionKeyPathTokens cannot be null");
+        List<String> propertiesToEncrypt =
+            this.includedPaths.stream().map(clientEncryptionIncludedPath -> clientEncryptionIncludedPath.getPath().substring(1)).collect(Collectors.toList());
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ClientEncryptionPolicy.java:94]: nice usage of java8 stream API. :-)

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncDatabase.java [hunk 2]
```diff
@@ -82,6 +82,13 @@ public CosmosAsyncClientEncryptionKey getClientEncryptionKey(String id) {
 
         EncryptionKeyStoreProvider encryptionKeyStoreProvider =
             this.cosmosEncryptionAsyncClient.getEncryptionKeyStoreProvider();
+
+        if (!encryptionKeyStoreProvider.getProviderName().equals(encryptionKeyWrapMetadata.getType())) {
+            throw new IllegalArgumentException("The EncryptionKeyWrapMetadata Type value does not match with the " +
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncDatabase.java:87]: in similar scenario, does non Cosmos Exception type reach public surface area in the DotNet SDK?

**Review by moderakh — COMMENTED** [review]
@simplynaveen20 in similar scenario, does non Cosmos Exception type reach public surface area in the DotNet SDK?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncDatabase.java (hunk 2 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncDatabase.java:87]: Yes .NET has ArgumentException as well. To me this is correct and it should be non cosmos as it is validation on argument passed by user.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ClientEncryptionPolicy.java [hunk 3]
```diff
@@ -76,6 +79,33 @@ public ClientEncryptionPolicy() {
         return this.includedPaths;
     }
 
+    /**
+     * Version of the client encryption policy definition.
+     * @return policyFormatVersion
+     */
+    @Beta(value = Beta.SinceVersion.V4_15_0, warningText = Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ClientEncryptionPolicy.java:86]: Should be changed to 4_16_0 version.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/EncryptionKeyWrapMetadata.java [hunk 4]
```diff
@@ -38,17 +38,15 @@ public EncryptionKeyWrapMetadata(EncryptionKeyWrapMetadata source) {
...
     @Beta(value = Beta.SinceVersion.V4_14_0, warningText = Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
-    public EncryptionKeyWrapMetadata(String name, String value) {
-        this("custom", name, value);
-    }
-
-    private EncryptionKeyWrapMetadata(String type, String name, String value) {
+    public EncryptionKeyWrapMetadata(String type, String name, String value) {
         Preconditions.checkNotNull(type, "type is null");
         Preconditions.checkNotNull(value, "value is null");
+        Preconditions.checkNotNull(name, "value is null");
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/EncryptionKeyWrapMetadata.java:49]: Incorrect message here - should be `name is null`

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/EncryptionKeyWrapMetadata.java [hunk 5]
```diff
@@ -90,6 +88,18 @@ public String getName() {
...
     }
 
+    /**
+     * Serialized form of metadata.
+     * Note: This value is saved in the Cosmos DB service.
+     * implementors of derived implementations should ensure that this does not have (private) key material or
+     * credential information.
+     * @return type of metadata.
+     */
+    @Beta(value = Beta.SinceVersion.V4_15_0, warningText = Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/EncryptionKeyWrapMetadata.java:98]: Beta 4_16_0

**Review by kushagraThapar — CHANGES_REQUESTED** [review]

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/EncryptionKeyWrapMetadata.java (hunk 5 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/EncryptionKeyWrapMetadata.java]: done

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/EncryptionKeyWrapMetadata.java (hunk 4 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/EncryptionKeyWrapMetadata.java]: corrected it

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ClientEncryptionPolicy.java (hunk 3 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ClientEncryptionPolicy.java]: done

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks @simplynaveen20

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncDatabase.java (hunk 2 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncDatabase.java:87]: how would it work on the async API? do we do validation upfront? or do we return a `Mono.of(IllegalArgException)`

**Review by moderakh — APPROVED** [review]
LGTM. except the question on the async API and how validation and throwing `IllegalArgException` from async api works.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncDatabase.java (hunk 2 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncDatabase.java:87]: This validation is om async API itself which is doing validation upfront .

**Pull request merged**